File: Rust/Week1/Day4/day4/src/bin/ex3.rs
fn main() {
    let party = vec![
        ("Astarion", 32, 40),
        ("Shadowheart", 20, 45),
        ("Karlach", 55, 60),
        ("Wyll", 0, 50),
        ("Lae'zel", 47, 50)
    ];
    
    let mut message = String::from("");
    for character in party.iter() {
        let health_percentage = 100.0*(character.1 as f32)/(character.2 as f32);
        let character_status = match health_percentage {
            75.0 ..= 100.0 => "Healthy",
            50.0 .. 75.0 => "Hurt",
            25.0 .. 50.0 => "Wounded",
            1.0 .. 25.0 => "Critical",
            _ => "Dead"
        };
        message.push_str("{charcter.0}: character.1/character.2 ({health_percentage:.2}%) -> {character_status}\n");
        //Question: Isn't there a way to make the line above work? there is no formatting. I wanted to reduce the number of print calls to one by apending the messages.
        print!("{}: {}/{} ({:.2}%) -> {}\n", character.0, character.1, character.2, health_percentage, character_status)
        //Question: The following case seems to be invalid "{character.0}" is there no simpler way to use the tupples in the string?
    }
    
    print!("{message}");
    //Question: The above is quite dumb and unintuitive, isn't there a better way to print a String?
}

File: CSharp/Week1/Day6/Day6/ex3.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Day6
{
    internal class ex3
    {
        static void Main()
        {
            var party = new (string name, int hp, int max)[]
            {
                ("Astarion", 32, 40),
                ("Shadowheart", 20, 45),
                ("Karlach", 55, 60),
                ("Wyll", 0, 50),
                ("Lae'zel", 47, 50)
            };

            foreach ((string name, int hp, int max) character in party)
            {
                float healthPercent = (100.0f * character.hp / character.max);
                String msg = String.Format("{0}: {1}/{2} ({3}/%) -> ", character.name, character.hp, character.max, healthPercent);
                switch (healthPercent)
                {
                    case >= 75:
                        {
                            msg += "Healthy";
                            break;
                        }
                    case >= 50:
                        {
                            msg += "Hurt";
                            break;
                        }
                    case >= 25:
                        {
                            msg += "Wounded";
                            break;
                        }
                    case > 0:
                        {
                            msg += "Critical";
                            break;
                        }
                    default:
                        {
                            msg += "Dead";
                            break;
                        }
                }
                Console.WriteLine(msg);
            };

            Console.WriteLine("Wounded characters:");
            var woundedPartyMembers = party.Where(c => c.hp <= c.max * 0.25f);
            foreach ((string name, int hp, int max) character in woundedPartyMembers)
            {
                Console.WriteLine("-{0}",character.name);
            }
        }
    }
}
